src/alvin0319/OffHand/OffHand.php

```diff
@@ -30,84 +30,88 @@
-		}
-	}
-
-	private function saveInventory(Player $player) : void{
-		$player->namedtag->setTag($this->getOffHandInventory($player)->getItemInOffHand()->nbtSerialize(-1, "OffHand"));
-		unset($this->inventories[$player->getName()]);
+        	$player = $event->getPlayer();
+        	$this->getOffhandInventory($player)->sendContents($player);
 	}
 
+	/**
+	 * @ignoreCancelled true
+	 * @priority MONITOR
+	 */
 	public function onDataPacketReceive(DataPacketReceiveEvent $event) : void{
 		$packet = $event->getPacket();
 		$player = $event->getPlayer();
-		if($packet instanceof MobEquipmentPacket){
-			if($packet->windowId === ContainerIds::OFFHAND){
-				$inv = $this->getOffHandInventory($player);
-				if($inv instanceof OffHandInventory){
-					$inv->setItemInOffHand($packet->item);
-					$event->setCancelled();
-				}
-			}
-		}
+        	if($packet instanceof MobEquipmentPacket and $packet->windowId == ContainerIds::OFFHAND){
+            		$offhand = $this->getOffhandInventory($player);
+            		if(!$offhand->getItem(0)->equalsExact($packet->item)){
+                		$offhand->sendContents($player);
+                		$event->setCancelled();
+                		return;
+            		}
+            		$offhand->setItem(0, $packet->item);
+        	}
 	}
-
+	/**
+	 * @ignoreCancelled true
+	 * @priority MONITOR
+	 */
 	public function onDataPacketSend(DataPacketSendEvent $event) : void{
-		$packet = $event->getPacket();
-		if($packet instanceof AddPlayerPacket){
-			if(($player = $this->getServer()->getPlayerExact($packet->username)) !== null){
-				$this->getOffHandInventory($event->getPlayer())->sendMobEquipmentPacket($player);
-				$this->getOffHandInventory($event->getPlayer())->sendMobEquipmentPacket($event->getPlayer());
-			}
-		}
+        	$packet = $event->getPacket();
+        	$player = $event->getPlayer();
+        	if($packet instanceof AddPlayerPacket){
+            		$this->getOffhandInventory($player->getServer()->getPlayerExact($packet->username) ?? $player)->sendItem([$player]);
+        	}
 	}
-
+	/**
+	 * @priority MONITOR
+	 */
 	public function onDeath(PlayerDeathEvent $event) : void{
-		$player = $event->getPlayer();
-		$drops = $event->getDrops();
-		if(!$event->getKeepInventory()){
-			$drops = array_merge($drops, $this->getOffHandInventory($player)->getContents(false));
-			$event->setDrops($drops);
-			$this->getOffHandInventory($player)->clearAll();
+        	$player = $event->getPlayer();
+        	$offhand = $this->getOffhandInventory($player);
+        	if(!$event->getKeepInventory() and !empty($event->getDrops())){
+            		$event->setDrops(array_merge($event->getDrops(), array_filter([$offhand->getItem(0)], function(Item $item): bool{
+                		return !$item->hasEnchantment(Enchantment::VANISHING);
+            		})));
+            		$offhand->clearAll();
 		}
+        }
+
+	public function getOffHandInventory(Human $player): OffHandInventory{
+        	static $players = [];
```

Comment:
Don't use static property for this function. this will cause memory leak.

Reply:
Oh didn’t know that

Reply:
Wait but that property should be in the class itself and can also be a private static without memory leak, right?

Reply:
In the Player handleMobEquipment you’ll see the same as this, so this is likely handled the same way as pmmp does with the Item In Hand, and basically it works perfectly